### ISSUE TITLE
[Feature request] Guide/example/support for "tree shaking" - e.g. a build that only includes a particular tokenizer, or a specific pipeline

### ISSUE THREAD

**Issue by josephrocca** [huggingface/transformers.js#407]
Referencing Xenova's comment [here](https://github.com/xenova/transformers.js/pull/389#issuecomment-1819510837), this may already be possible via existing build tools. If anyone has investigated this, it would be great if there were an example in the docs for those of us who aren't as savvy with this sort of thing.

IIRC, there are some types of API designs which aren't amenable to tree shaking (just hunted down [this video](https://youtu.be/lsd2-TCgHEs?si=AjLJoMLhEHiG9GaX&t=133)), although I'm not sure how true that is with the progress that has been made with typescript/jsdoc static analysis, and (perhaps naively) it seems like an "optimistic treeshaking" would be fine for 99% of use cases anyway. But I have basically zero experience here so I'm mentioning this just in case it's relevant to any API design choices in the future. Hopefully someone with experience here can comment.

**Comment by aadito123** [comment]
@xenova  Sorry for the ping on an old thread. Unfortunately, the current code architecture is very unfriendly to tree-shaking through regular static analysis. This is mostly due to the (heavy) reliance on reflection and lookup maps. For example:
The `BertModel` just inherits from the base class `PreTrainedModel`. However, the `PreTrainedModel` references all of its subclasses via the `MODEL_CLASS_TO_NAME_MAPPING`. This prohibits build tools (admittedly only test esbuild and rollup) from being able to treeshake them out. JS classes, in general, tend to be hard to treeshake with. 

Would you be open to the refactoring (and likely API changes) required for treeshaking? I would love to help work on this feature.

Thanks!

**Comment by BLucky-gh** [comment]
I know this is yet another instance of necroposting but I've noticed that the sole export from the package is the `transformers.js` file which is just one big barrel export, and barrel files are notoriously bad for tree-shaking

if the component parts of the package (for example just `transformers.js`) were defined as package exports in `package.json` individually (in addition to the barrel file, for backwards compatibility), that would likely help with treeshaking, though that would require changing the build process to emit the intermediate files instead of bundling it all into one file

**Comment by BLucky-gh** [comment]
small update: tried it on my own: shaved off only about 200kB (min+gzipped) when importing only `tokenizers.js` and the majority of it is still the dependency on `onnxruntime-web` since the `Tensor` class imported from `tensor.js` uses the `Tensor` class from `onnxruntime-web`, so it would require more work than just individual non-barrel exports to be able to tree-shake more of it

**Comment by xenova** [comment]
A separate tokenizers library is now available :)
https://www.npmjs.com/package/@huggingface/tokenizers 🔥 

Repo is public too: https://github.com/huggingface/tokenizers.js

We look forward to seeing what you build with it!

**Comment by rxliuli** [comment]
Just wanted to ask, would you consider accepting a PR to change the single-file bundle to multiple files? Currently, even with split chunks, the single file is still 848kb after building, which is really too large. Vite shows a warning every time it builds.

```sh
dist/assets/huggingface-transformers-DkRbcDfl.js       848.72 kB │ gzip:   216.75 kB
[plugin builtin:vite-reporter] 
(!) Some chunks are larger than 500 kB after minification. Consider:
- Using dynamic import() to code-split the application
- Use build.rollupOptions.output.manualChunks to improve chunking: https://rollupjs.org/configuration-options/#output-manualchunks
- Adjust chunk size limit for this warning via build.chunkSizeWarningLimit.
- ```

**Comment by nico-martin** [comment]
Hi @rxliuli,
Thank you very much for your offer. However, I would like to wait for v4 first. We have made a few fundamental adjustments for the next major version, and others are still in the planning stage for v4.

1. Trokenizer will be outsourced and imported from the new tokenizers.js package.
2. The build process is now esbuild wich resulted in slightly smaller filesizes (whereby the main reason for the change was the speed).
3. We are currently experimenting with path exports or subpackages for transformers.js-related packages (framework abstractions, etc.), which is why the npm setup will change quite a bit again.

As I see it, most of the code is shared logic (cannot be tree-shaken because it is used by every model/task) or tokenizers (refactoring in progress). The actual model/task implementations are hardly worth mentioning.

To sum it up, improvements are always welcome, but since right now there are a couple of big refactorings going on I would suggest to wait for a bit.
